probe-rs/src/architecture/riscv/communication_interface.rs

```diff
@@ -0,0 +1,848 @@
+                Ok(_) => break,
+                Err(RiscvError::DmiTransfer(DmiOperationStatus::RequestInProgress)) => {
+                    // Operation still in progress, reset dmi status and try again.
+                    self.dmi_reset()?;
+                }
+                Err(e) => return Err(e),
+            }
+
+            if start_time.elapsed() > retry_duration {
+                return Err(RiscvError::Timeout);
+            }
+        }
+
+        let response;
+
+        loop {
+            // Perform the actual read
+            match self.dmi_register_access(0, 0, DmiOperation::NoOp) {
+                Ok(r) => {
+                    response = r;
+                    break;
+                }
+                Err(RiscvError::DmiTransfer(DmiOperationStatus::RequestInProgress)) => {
+                    // Operation still in progress, reset dmi status and try again.
+                    self.dmi_reset()?;
+                }
+                Err(e) => return Err(e),
+            }
+
+            if start_time.elapsed() > retry_duration {
+                return Err(RiscvError::Timeout);
+            }
+        }
+
+        //let response = self.dmi_register_access(0, 0, DmiOperation::Read)?;
+
+        log::debug!(
+            "Read DM register '{}' at {:#010x} = {:#010x}",
+            R::NAME,
+            R::ADDRESS,
+            response
+        );
+
+        Ok(response.into())
+    }
+
+    pub(crate) fn write_dm_register<R: DebugRegister>(
+        &mut self,
+        register: R,
+    ) -> Result<(), RiscvError> {
+        // write write command to dmi register
+
+        let data = register.into();
+
+        log::debug!(
+            "Write DM register '{}' at {:#010x} = {:#010x}",
+            R::NAME,
+            R::ADDRESS,
+            data
+        );
+
+        self.dmi_register_access(R::ADDRESS as u64, data, DmiOperation::Write)?;
+
+        Ok(())
+    }
+
+    /// Perfrom memory read from a single location using the program buffer.
+    /// Only reads up to a width of 32 bits are currently supported.
+    /// For widths smaller than u32, the higher bits have to be discarded manually.
+    fn perform_memory_read(&mut self, address: u32, width: u8) -> Result<u32, RiscvError> {
+        // assemble
+        //  lb s1, 0(s0)
+
+        // Backup registers s0 and s1
+        let s0 = self.abstract_cmd_register_read(0x1008)?;
+
+        //let o = 0; // offset = 0
+        //let b = 9; // base register -> s0
+        //let w = 0; // width
+        //let d = 9; // dest register -> s0
+        //let l = 0b11;
+
+        let mut lw_command: u32 = 0b000000000000_01000_000_01000_0000011;
+
+        // verify the width is supported
+        // 0 ==  8 bit
+        // 1 == 16 bit
+        // 2 == 32 bit
+        assert!(width < 3, "Width larger than 3 not supported yet");
+
+        lw_command |= (width as u32) << 12;
+
+        let ebreak_cmd = 0b000000000001_00000_000_00000_1110011;
```

Comment:
Consider checking for impebreak capability in dmstatus, also check for program buffer size.

There can actually be a debug module, which has only one Progbuf slot, and uses the implicit ebreak to return to normal execution.

Reply:
I've added a new issue #151 for this. I'd like to merge this PR quite soon, so that people can also work on ARM support for JLink, so I will handle this in another PR. I agree that this should be checked, and supported.